koala/Spreadsheet.py

```diff
@@ -175,7 +175,12 @@ def gen_graph(self, outputs=[], inputs=[]):
         # undirected = networkx.Graph(G)
         # print "Number of connected components %s", str(number_connected_components(undirected))
 
-        self.build_spreadsheet(G, cellmap, self.named_ranges, pointers = self.pointers, outputs = outputs, inputs = inputs, debug = self.debug)
+        if inputs == [] and outputs == []:
```

Comment:
@Brad-eki, thanks for fixing the bug that I introduced. This indeed adds the old behaviour back, but I don't like the fact that it is indirectly based upon two inputs. Can we not make it explicit by adding an argument to gen_graph whether or not to return this object? The default can be yes (to restore the old behaviour). Internally, gen_graph() is only used in one place (line 65 in the same file) so we should specify there the new argument such that it won't return the new object. The result is the same, but I think the code will be clearer.

Reply:
I agree the fix I have proposed is well less than ideal. In the same breath I'm less than convinced adding a flag would start to address the base concerns.

From the current __init__, the Spreadsheet object (and essentially Koala more generally) is required in at least two use cases;
- "loading"/graphing an entire workbook
- "loading"/graphing a subset of the workbook through use of seed cells

I accept most people are likely to want the first use case but my main use case is the second one. The way I have found to achieve the second behavior is to load the entire workbook and then call gen_graph a number of times giving specific input and output cells to pick and choose the different subsets I'm interested in. I'm not really convinced this is working correctly either.

The core of all the graph creation and updating is ast/__init__.graph_from_seeds(seeds, cell_source).

The comments for graph_from_seeds note:
    The graph is created when the cell_source is an instance of ExcelCompiler
    The graph is updated when the cell_source is an instance of Spreadsheet

How do we tell it's a spreadsheet?:
    if hasattr(cell_source, 'G'): # ~ cell_source is a Spreadsheet

And currently all calls to graph_from_seeds are both from Spreadsheet and they all pass self as the cell_source. We are not exercising the "ExcelCompiler"/create branch.

So I think we are on the cusp of a more comprehensive change. One where we clean up the AST code. And potentially have an interface to AST where there is each a separate method for creating and updating the graph. Or potentially some different language which describes more accurately the possible operations on a graph.

This would then put us in a good position to improve the interface to Spreadsheet where it is then possible to directly operate using the first or second use cases, and add other cases as they arise.

The ast code is messy and is dealing with some difficult concepts. I tried to learn Abstract Syntax Trees earlier this year to get my head around how Koala operates and they are, appropriately, rather abstract.

To my mind the ast code is using poor variable names eg; single letter or short codes like c1, e, t and G. Only a personal preference but I like variable names that are expressive and easily searchable.

Ast as it stands is very confusing. One example comment:
  # Whats the difference between subgraph() and make_subgraph() ?
When you read the methods they are rather similar.

I don't think we can solve this elegantly with my proposed change, or with the use of a flag. It can be demonstrated that the "create" decision tree is simply not being used (and yet to be determined if it even needs to be). At the same time I can't use Koala post issue #188 unless we;
- implement your suggestion
- implement my suggestion
- fix AST (or, maybe, adopt someone else's) and refactor Spreadsheet.

Reply:
maybe we can explore this: https://pypi.org/project/iAST/

Reply:
To be fair, there isn't much to Koala outside the AST.

So getting a good implementation for ASTs would allow us to adapt the desired functionality reasonably quickly.

Reply:
https://pypi.org/project/astutils/

Reply:
A great place to find an interface for an AST is the Python language AST..

https://docs.python.org/3/library/ast.html

Obviously that one is written to interpret the Python language but nothing says we can't copy the design.

Reply:
And there's an AST over in the closest "competition": Formulas

https://github.com/vinci1it2000/formulas/blob/master/formulas/builder.py

Our niche and requirements are different to Formulas, but nonetheless they have a really clean implementation we might use as inspiration.

Reply:
@Brad-eki, yes I also experienced some issues in going through the AST code. I must admit I still don't understand it completely. I understand your comments and I see that we should move forward to restore the old behaviour. I also see your point to say that my proposal isn't perfect either. For now I will merge your changes as such and in a next step we can (step by step preferably) start to change the AST part. Once we get back here, we can use the same principle of deprecated messages as I used elsewhere to steer people towards other functions.